clients/src/main/java/org/apache/kafka/clients/ClientTelemetryState.java

```diff
@@ -0,0 +1,167 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients;
+
+import org.apache.kafka.common.errors.IllegalClientTelemetryStateException;
+import org.apache.kafka.common.utils.Utils;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * State that helps determine where client exists in the telemetry state i.e. subscribe->wait->push loop.
+ */
+public enum ClientTelemetryState {
+
+    /**
+     * Client needs subscription from the broker.
+     */
+    SUBSCRIPTION_NEEDED,
+
+    /**
+     * Network I/O is in progress to retrieve subscription.
+     */
+    SUBSCRIPTION_IN_PROGRESS,
+
+    /**
+     * Awaiting telemetry interval for pushing metrics to broker.
+     */
+    PUSH_NEEDED,
+
+    /**
+     * Network I/O in progress for pushing metrics payload.
+     */
+    PUSH_IN_PROGRESS,
+
+    /**
+     * Need to push the terminal metrics payload.
+     */
+    TERMINATING_PUSH_NEEDED,
+
+    /**
+     * Network I/O in progress for pushing terminal metrics payload.
+     */
+    TERMINATING_PUSH_IN_PROGRESS,
+
+    /**
+     * No more work should be performed, telemetry client terminated.
+     */
+    TERMINATED;
+
+    private final static Map<ClientTelemetryState, List<ClientTelemetryState>> VALID_NEXT_STATES = new EnumMap<>(ClientTelemetryState.class);
+
+    static {
+        /*
+         If clients needs a subscription, then issue telemetry API to fetch subscription from broker.
+
+         However, it's still possible that client doesn't get very far before terminating.
+        */
+        VALID_NEXT_STATES.put(
+            SUBSCRIPTION_NEEDED, Arrays.asList(SUBSCRIPTION_IN_PROGRESS, TERMINATED));
+
+        /*
+         If client is finished waiting for subscription, then client is ready to push the telemetry.
+         But, it's possible that no telemetry metrics are requested, hence client should go back to
+         subscription needed state i.e. requesting the next updated subscription.
+
+         However, it's still possible that client doesn't get very far before terminating.
+        */
+        VALID_NEXT_STATES.put(SUBSCRIPTION_IN_PROGRESS, Arrays.asList(PUSH_NEEDED,
+            SUBSCRIPTION_NEEDED, TERMINATING_PUSH_NEEDED, TERMINATED));
```

Comment:
Why do we allow both `TERMINATING_PUSH_NEEDED` and `TERMINATED`? How could we skip `TERMINATING_PUSH_NEEDED` and go to `TERMINATED` directly? Similar for other transitions below.

Reply:
Agree ideally that should be the transitioning order but as the KIP says: `or it can choose to abandon sending a final metrics push.`

i.e. if signal received to terminate the client and the client was earlier in  `SUBSCRIPTION_IN_PROGRESS` where some error was received then client can decide to abandon the last push and move to terminated state.

Please let me know if I misunderstood this.

```
In the event that the client's metric subscription has changed and the final metrics push fails with error code UNKNOWN_SUBSCRIPTION_ID , the terminating client can choose to obtain a new subscription ID by sending a GetTelemetrySubscriptionsRequest and then immediately sending a PushTelemetryRequest with the Terminating flag set to true, or it can choose to abandon sending a final metrics push.

Reply:
There also may be an issue of respecting timeouts for the close of the client. If there isn't enough time to send the final push telemetry, head straight for the exit.

Reply:
Thanks for explaining. My thought was that we would still transit to `TERMINATING_PUSH_NEEDED`, but potentially "skip" this state, and do an immediate second transition to `TERMINATED` -- but if't easier to transit to `TERMINATED` directly, also fine with me. (Just want to understand what's going on.)